### ISSUE TITLE
Race Condition on loading index from multiple non workspace locking threads

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-eclipse#1494]
there is a Race Condition on loading index from multiple non workspace locking threads
```
CustomEMFBasedPersister.load() line: 87	
AbstractBuilderState$1.run(IProgressMonitor) line: 78	
Workspace.run(ICoreRunnable, ISchedulingRule, int, IProgressMonitor) line: 2292	
Workspace.run(IWorkspaceRunnable, ISchedulingRule, int, IProgressMonitor) line: 2317	
CustomClusteringBuilderState(AbstractBuilderState).load() line: 92	
CustomClusteringBuilderState(AbstractBuilderState).ensureLoaded() line: 107	
CustomClusteringBuilderState(AbstractBuilderState).isEmpty() line: 254	
SomeOtherUtil.lambda$0(IProgressMonitor) line: 33	
1321697288.run(IProgressMonitor) line: not available	
Job$2.run(IProgressMonitor) line: 185	
Worker.run() line: 63	
```

```
CustomBasedPersister.load() line: 87	
AbstractBuilderState$1.run(IProgressMonitor) line: 79	
Workspace.run(ICoreRunnable, ISchedulingRule, int, IProgressMonitor) line: 2292	
Workspace.run(IWorkspaceRunnable, ISchedulingRule, int, IProgressMonitor) line: 2317	
CustomClusteringBuilderState(AbstractBuilderState).load() line: 92	
CustomClusteringBuilderState(AbstractBuilderState).ensureLoaded() line: 107	
CustomClusteringBuilderState(AbstractBuilderState).isEmpty() line: 254	
DirtyStateAwareResourceDescriptions.isEmpty() line: 143	
SomeUtil.initialize() line: 319	
ExecutionHierarchyView.createViewer(Composite) line: 467	
ExecutionHierarchyView.createPartControl(Composite) line: 303	
CompatibilityView(CompatibilityPart).createPartControl(IWorkbenchPart, Composite) line: 154	
```

as you can see `(AbstractBuilderState).load() line: 92` both threads are in the `NO` case and call `workspace.run` with no scheduling rule. i assume we have to pass `ResourcePlugin.getWorkspace().getRoot()` there.

what do you thing @szarnekow? cc @iloveeclipse

**Comment by szarnekow** [comment]
Please see also the comment in `org.eclipse.xtext.ui.resource.Storage2UriMapperJavaImpl.initializeCache(boolean)` I don't think it's possible to use a scheduling rule here.
We could synchronize org.eclipse.xtext.builder.builderState.EMFBasedPersister.load() on the class object, e.g.
```
 .. load() {
  synchronized(PersistedStateProvider.class) {
    .. current impl here
  }
}
```

**Comment by iloveeclipse** [comment]
Or we could create IJobManager.newLock() in the provider and use this lock - this could avoid deadlocking because deadlocks on ILock's can be resolved by platform.

**Comment by cdietrich** [comment]
locking in the Persister does not work as the second one will return empty list after called.
=> we would have to move the lock again into the runnable.
nut i fear that one will introduce some deadlocks again

**Comment by cdietrich** [comment]
Fixed in Xtext 2.23
